feat: add support for preserving and labeling intermediate stage images

ezopezo opened the pull request:
This adds support for preserving and labeling intermediate stage images (including final image) in multi-stage builds. `--save-stages` preserves only the final image from each intermediate stage, not every instruction layer (as `--layers` flag). This also keeps the final image's layer count same as in regular builds (with no additional args). `--stage-labels` adds label metadata containing base image and alias from Containerfile at the beginning of the stage.


New flags:
  - `--save-stages`: save intermediate stage final images instead of removing them.
        Build without added `--layers`, intermediate images are expected to accumulate on each build (no cache reuse).
        Build with added `--layers`, enables layer cache reuse for subsequent builds.

  - `--stage-labels`: adds metadata labels to intermediate stage images and final image:
    * `io.buildah.stage.name`: the stage name (alias or stage index if alias not specified)
    * `io.buildah.stage.base`: the base image (external image reference - pullpsec or
          parent stage image ID if its base is another stage)
    Requires `--save-stages`.
    Cache reuse depends of if first or second build has or has not `–stage-labels` argument or if alias/pullspec changed between builds in Containerfile

The implementation includes:
  - Validation that `--stage-labels` requires `--save-stages`
  - Detection when a stage uses another intermediate stage as base (and labeling the stage accordingly)
  - Stage labels are added to Containerfile via parse tree injection as LABEL at the beginning of the stage including final
  - Various test coverage for saving, labeling, and caching scenarios
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

**General use**: This functionality is useful for identification and debugging intermediate stage images in multi-stage builds.

**Specific need**: Identifying the content copied from intermediate stages in multi-stage builds into the final image is a hard requirement for supporting Contextual SBOM - an SBOM that understands the origin of each package.
While intermediate images can be extracted using the `--layers` option, this approach has several issues for our use case:
- Intermediate stage images are unlabeled, making it difficult to determine which image corresponds to which build stage - especially when the Containerfile reuses the same pullspec across multiple stages (pullspecs does not have to be unique, aliases must be).
- All instructions from all intermediate stages appear in the cache (visible via `buildah images --all`), which introduces unnecessary noise for our purposes.
- rootfs.diff_ids are not squashed in final stage: the final-stage image ends up containing diff IDs for every instruction in the final stage (due to the instructions reuse also for final stage). However, we need the final build image to resemble a regular build (without `--layers`), meaning:
  - it should contain the diff IDs inherited from the base image, and
  - exactly one diff ID representing the squashed final-stage instructions.

Related repositories: 
[konflux](https://github.com/konflux-ci/build-definitions) (uses mobster for SBOM generation), 
[mobster](https://github.com/konflux-ci/build-definitions) (implements contextual SBOM functionality requiring this change), 
[capo](https://github.com/konflux-ci/capo) (wraps builder content identification functionality for mobster), 
Contact person: emravec (RedHat) / @ezopezo (Github)

#### How to verify it

Run any multistage build with intermediate stage, with implemented arguments. Resulting intermediate images should be saved and correctly labeled. Example:
`buildah build --save-stages --stage-labels -t test:0.1 .`

#### Which issue(s) this PR fixes:

Fixes: #6257
Internal Jira: https://issues.redhat.com/browse/ISV-6122

#### Does this PR introduce a user-facing change?

```release-note
Add `--save-stages` and `--stage-labels` flags for preserving and labeling intermediate stage images in multi-stage builds.
```

ezopezo commented:
/retest

nalind commented:
/ok-to-test

mtrmac commented:
@flouthoc if you have time

TomSweeneyRedHat commented on docs/buildah-build.1.md:
```diff
@@ -136,6 +143,19 @@ the intermediate image is stored in the image itself. Buildah's approach is simi
+**--cache-stages** *bool-value*
+
+Preserve intermediate stage images instead of removing them after the build completes
+(Default is `false`). By default, buildah removes intermediate stage images to save space.
```
```suggestion
(Default is `false`). By default, Buildah removes intermediate stage images to save space.
```

TomSweeneyRedHat commented on docs/buildah-build.1.md:
```diff
@@ -1385,6 +1420,10 @@ buildah build -v /var/lib/dnf:/var/lib/dnf:O -t imageName .
 
 buildah build --layers -t imageName .
 
+buildah build --cache-stages --stage-labels -t imageName .
```
just to show the bool values in play, perhaps
```suggestion
buildah build --cache-stages false --stage-labels false -t imageName .
```

TomSweeneyRedHat commented on docs/buildah-build.1.md:
```diff
@@ -1443,6 +1482,21 @@ buildah build --output type=tar,dest=out.tar .
+Find an intermediate image for a specific stage name:
+
+buildah images --filter "label=io.buildah.stage.name=builder"
+
```
This is great!  TY!

TomSweeneyRedHat commented on pkg/cli/common.go:
```diff
@@ -26,8 +26,10 @@ import (
-	Layers  bool
+	ForceRm     bool
+	Layers      bool
+	CacheStages bool
```
More OCD alpha changes, please put this at line 28.5

TomSweeneyRedHat reviewed (commented):
First pass LGTM, just a couple of nitty things here and there.

@nalind or @Luap99 PTAL

Also, have you tried vendoring Buildah into Podman with these changes yet?


@ddarrah should we have a feature/epic for this in RHEL 9.8/10.2?

nalind commented on imagebuildah/stage_executor.go:
```diff
@@ -1837,6 +1843,27 @@ func (s *stageExecutor) execute(ctx context.Context, base string) (imgID string,
+			logrus.Debugf("Committing intermediate stage %s (index %d) for --cache-stages", s.name, s.index)
+			createdBy := fmt.Sprintf("/bin/sh -c #(nop) STAGE %s", s.name)
+			// Commit the stage without squashing, using empty output name (intermediate image)
+			imgID, commitResults, err = s.commit(ctx, createdBy, false, "", false, false)
```
Why set `emptyLayer` as `false` here?

ezopezo replied:
You re right - it should not be false. I updated it such way, if stage was already committed, we only add metadata labels without creating a duplicate empty layer. Let me know, if it is okay!

nalind reviewed (commented):
Overall I think I understand the goals here.  The new flags interact with `--layers` in confusing ways, though - the build ID recorded for `--build-id-file` won't match the image whose ID we return if the build is a cache hit, but we may quietly commit new versions of the final product of intermediate stages.  I'm not really sure what the intended result is there.

ezopezo commented on define/build.go:
```diff
@@ -278,6 +280,12 @@ type BuildOptions struct {
+	CacheStages bool
+	// StageLabels tells the builder to add metadata labels to intermediate stage images for easier recognition.
+	// These labels include stage name, base image, build ID, and parent stage name (when a stage uses another
+	// intermediate stage as its base, i.e., transitive aliases). This option requires CacheStages to be enabled.
```
Fair point - you're right that “transitive alias” isn't established jargon. I started using it internally as shorthand for multi-stage builds where one named stage is used as the base for another, and it unfortunately leaked here. I think I picked it up from an article or docs at some point, but I can't find a reference now - which indeed makes it sound like it was revealed to me in a dream 🙂
I'll happily rephrase all occurrences using some more standard terminology!

ezopezo replied:
Done :)

ezopezo commented:
> Overall I think I understand the goals here. The new flags interact with `--layers` in confusing ways, though - the build ID recorded for `--build-id-file` won't match the image whose ID we return if the build is a cache hit, but we may quietly commit new versions of the final product of intermediate stages. I'm not really sure what the intended result is there.

Does it make a sense to add an mutual exclusion condition that --layers and --cache-stages cannot be used together (as serve fundamentally different purposes)? I want to keep this functionality as isolated as possible from others to avoid unnecessary complexities like this.

ezopezo commented:
> > @nalind Also I almost forgot to ask - how I can effectively test if current changes are compatible with podman? (I believe this is what @TomSweeneyRedHat asked before as well):
> 
> Since we're on `main`, check out podman's `main` branch, and use `go mod edit -replace github.com/containers/buildah=github.com/ezopezo/buildah@478bf2159f6ffe878c5162ed1d83c69ae3e6a6d6 && go mod tidy && go mod vendor && make` to build. (If your branch name was a valid version string, you could use its name directly instead of having to look up the commit ID.) At that point you can either attempt to test locally with `make .install.ginkgo localintegration` and related targets, or commit everything to a branch, push it to a podman fork, and open a PR to have podman's CI do it, taking care to mark it as not intended for merging.

@nalind Thank you I did it according your guide and created [this testing PR](https://github.com/containers/podman/pull/27964). I think it is failing because of the docs mismatch and exceeding some limit of the binary size (which could be overridden by label I think) and also I think I need from maintainers approval for running all of the jobs, could you help me with this please?

nalind commented:
I've given it a kick.

nalind commented on tests/bud.bats:
```diff
@@ -9542,3 +9542,718 @@ _EOF
+  # Second build should reuse the layers created by first build (cache hit)
+  run_buildah build $WITH_POLICY_JSON --layers --save-stages --stage-labels -t ${target}-second -f $BUDFILES/save-stages/Dockerfile.single-build-stage $BUDFILES/save-stages
+
+  expect_output --substring "Using cache"
```
If we're expecting cache hits at every step, I'd expect verifying that the image ID for the second build's final output matches that of the first to be a step in that.

nalind commented on tests/bud.bats:
```diff
@@ -9542,3 +9542,718 @@ _EOF
+
+  # Get intermediate stage image (RUN layer) and COPY layer from second build
+  second_run_layer=$(echo "$intermediate_images_second" | awk 'NR==1')
+  second_copy_layer=$(echo "$intermediate_images_second" | awk 'NR==3')
```
The order of values here is not guaranteed.

nalind commented on imagebuildah/executor.go:
```diff
@@ -604,6 +610,13 @@ func (b *executor) buildStage(ctx context.Context, cleanupStages map[int]*stageE
+	// Create stage labels for intermediate stage images
+	// Skip if stage has no instructions
+	if b.saveStages && b.stageLabels && stageIndex < len(stages)-1 && len(stage.Node.Children) > 0 {
+		labelLine := b.buildStageLabelLine(&stage, base, stages, stageIndex)
```
```suggestion
		labelLine := b.buildStageLabelLine(&stage, base, stages[:stageIndex])
```
1 of 3, so that the index of the current stage doesn't need to be passed in.

nalind reviewed (changes requested):
Possible issues when `--jobs` is used with values greater than 1, and questions around labels in the final stage and in stages that inherit from other stages in multi-layer builds.

ezopezo commented on imagebuildah/executor.go:
```diff
@@ -692,6 +713,32 @@ func markDependencyStagesForTarget(dependencyMap map[string]*stageDependencyInfo
+		labelLine += fmt.Sprintf(" %q=%q", "io.buildah.stage.base", base)
+	}
+	return labelLine
+}
```
Removed the conditions!

nalind reviewed (commented):
Some questions about the tests.

ezopezo committed:
feat: add support for saving and labeling intermediate stages

This adds support for saving and labeling all of the intermediate stage images
(including final image) in the multi-stage builds.
--save-stages preserves only the final image from each intermediate stage,
not every instruction layer (as --layers flag). This also keeps the final image's
layer count same as in regular builds (with no additional args).
--stage-labels adds label metadata containing base image and alias
from Containerfile at the beginning of the stage.

New flags:
  - --save-stages: save intermediate stage final images instead of removing them.
    Build without added --layers, intermediate images are expected to accumulate on each build (no cache reuse).
    Build with added --layers, enables layer cache reuse for subsequent builds.

  - --stage-labels: add metadata labels to intermediate stage images and final image:
    * io.buildah.stage.name: the stage name (alias or stage index if alias not specified)
    * io.buildah.stage.base: the base image (external image reference - pullpsec or
      parent stage image ID if its base is another stage)
    Requires --save-stages.
    Cache reuse depends of if first or second build has or has not –stage-labels
    argument or if alias/pullspec changed between builds in Containerfile.

The implementation includes:
  - Validation that --stage-labels requires --save-stages
  - Detection when a stage uses another intermediate stage as base (and labeling the stage accordingly)
  - Stage labels are added to Containerfile via parse tree injection as LABEL at the beginning of the stage including final
  - Various test coverage for saving, labeling, and caching scenarios

This functionality is useful for debugging, exploring, and reusing intermediate stage images from multi-stage builds.

Signed-off-by: Erik Mravec <emravec@redhat.com>

ezopezo commented on tests/bud/save-stages/Dockerfile.arg-build-stages-and-chained-build-stages:
```diff
@@ -0,0 +1,13 @@
+FROM alpine
+ARG SECOND_STAGE=second-stage
+COPY --from=${SECOND_STAGE} /output.txt /app/output.txt
+RUN cat /app/output.txt
```
Sorry I wrote it wrongly, I meant; ARGS stated at the top of the _Containerfile_ are not resolved _inside of any stage_ (only in FROM instructions as pullspecs or aliases). But it makes a sense I guess, that stages have their own ARG scope. Nevermind it is not important, I don't want to tangle into this I just wanted to explain why I did this weird stuff before:
```
ARG SECOND_STAGE=second-stage
COPY --from=${SECOND_STAGE} ...
```
But it is not needed at all as I explained

Anyway, is the Containerfile and test okay as it is now?

nalind reviewed (approved):
LGTM
@containers/buildah-maintainers PTAL
